src/main/java/de/tum/in/www1/artemis/service/StudentExamService.java

```diff
@@ -202,6 +207,24 @@ else if (exercise instanceof ModelingExercise) {
             }
         }
 
+        // Only lock programming exercises when the student submitted early. Otherwise, the lock operations were already scheduled/executed.
+        if (ZonedDateTime.now().isBefore(examEndDate)) {
```

Comment:
can you please:
1) extract a method for the whole new code block
2) move the invocation of the new method after the line `studentExamRepository.save(studentExam);`
3) Put another try catch statement around the invocation of the new method? So that really no (even currently unthinkable) exception can lead to problems here?